### PR TITLE
Add stats source and new dashboard for metrics from backup archives

### PR DESCRIPTION
**Pull request by Lackshan**
After the creation of cbmstatparser (a tool to create a Prometheus TSDB containing stats from cbbackupmgr runs), the tools team decided that integrating this tool into promtimer (to visualise the aforementioned TSDB with Grafana) would provide the most utiity, as it would allow one tool to display all important statistics for a Couchbase Server instance. To this effect, a flag was added to promtimer that when used would make it run cbmstatparser to create a Prometheus TSDB, run a prometheus instance with this TSDB, and then visualise it with Grafana.

### CODE REVIEW

File: promtimer/cbstats.py [hunk 1]
```diff
@@ -448,6 +450,88 @@ def get_user_log(cluster, user, password):
...
+        super().__init__(prometheus_port)
+        self._short_name = short_name
+        self._prometheus_tsdb_path = prometheus_tsdb_path
+        self._config = None
+
+    def short_name(self):
+        return self._short_name
+
+    def host(self):
+        return '127.0.0.1'
```
Comment (jamesl33) [inline promtimer/cbstats.py:468]: Could use `@property` here to avoid having to use `bs.short_name()` couldn't we?

File: promtimer/cbstats.py [hunk 2]
```diff
@@ -448,6 +450,88 @@ def get_user_log(cluster, user, password):
...
+        stat_files = []
+        for file in os.listdir(cpu_stats_dir):
+            if path.isfile(path.join(cpu_stats_dir, file)) and file[0] != '.':
+                stat_files.append(file)
+
+        timestamps = []
+        for file in stat_files:
+            # The timestamp in the filename indicates when the backup started
+            timestamps.append(datetime.datetime.fromtimestamp(int(file.split('-')[2])))
+        timestamps.sort()
```
Comment (jamesl33) [inline promtimer/cbstats.py:505]: Minor, but would call `sort()` after the `len(timestamps) == 0` check

File: promtimer/cbstats.py [hunk 3]
```diff
@@ -448,6 +450,88 @@ def get_user_log(cluster, user, password):
...
+
+        if len(timestamps) == 0:
+            raise FileNotFoundError('No cpu stat files present in ' + cpu_stats_dir)
+
+        start_timestamp = timestamps[0] - datetime.timedelta(minutes=1) # This padding makes the graph look better
```
Comment (jamesl33) [inline promtimer/cbstats.py:510]: I'd add a method for the padding, then you only need to document in one place and updates in the future will be easier.

File: promtimer/cbstats.py [hunk 4]
```diff
@@ -448,6 +450,88 @@ def get_user_log(cluster, user, password):
...
+
+        last_stat_file = os.path.join(
+            cpu_stats_dir,
+            'repo-backup-' + str(int(time.mktime(timestamps[-1].timetuple())))
```
Comment (jamesl33) [inline promtimer/cbstats.py:514]: Are we asserting that the list has at least two elements, or are we allowing it to use the same timestamp, and the padding will just give us a "window" to look into?

File: promtimer/cbstats.py [hunk 5]
```diff
@@ -448,6 +450,88 @@ def get_user_log(cluster, user, password):
...
+        )
+
+        with open(last_stat_file, 'rb') as f:
+            # Only read last line of file:
+            try:
+                f.seek(-2, os.SEEK_END) # Skip EOF char at end of file
+                while f.read(1) != b'\n':
+                    f.seek(-2, os.SEEK_CUR)
+            except OSError:
+                f.seek(0)
```
Comment (jamesl33) [inline promtimer/cbstats.py:524]: This is all quite hard to follow as we've got some magic numbers here, I think we should extract it into a general `read_last_line` function, and document the magic numbers.

File: promtimer/promtimer.py [hunk 6]
```diff
@@ -147,7 +170,14 @@ def prepare_grafana(grafana_port,
     make_home_dashboard()
     make_data_sources(stats_sources)
     make_dashboards_yaml()
-    make_dashboards(stats_sources, buckets, min_time_string, max_time_string, refresh)
+    if not cbbackupmgr_stats_mode:
```
Comment (jamesl33) [inline promtimer/promtimer.py:173]: I'd drop the `not` here and flip the logic since you're doing `if/else` anyway.

File: README.md [hunk 7]
```diff
@@ -113,6 +113,23 @@ The Grafana dashboards page should open for you automatically. If not, navigate
...
+promtimer --stats-archive-path <path-to-backup-archive-or-zip>
+```
+
+The `path-to-backup-archive-or-zip` specifies the path to either a backup archive or to
+a cbbackupmgr collect zip file.
+
+The Grafana dashboards page should open for you automatically. If not, navigate
+to `localhost:13300/dashboards` in your browser and begin exploring the
+available dashboards.
+
```
Comment (jamesl33) [inline README.md:132]: Should we document that this only works if you have access to the private `backup` repository.

File: promtimer/cbstats.py (hunk 1 above)
Comment (Lackshan) [reply promtimer/cbstats.py:468]: I'll keep this in mind for the future, but I think with how little I call the getter, it's not going to make much of a difference 😛

File: promtimer/cbstats.py (hunk 4 above)
Comment (Lackshan) [reply promtimer/cbstats.py:514]: The latter, we're allowing it to use the same timestamp (in the unlikely event there's only one) and the padding creates a window.

File: promtimer/cbstats.py (hunk 5 above)
Comment (Lackshan) [reply promtimer/cbstats.py:524]: Agreed, have moved it to a seperate function

File: promtimer/promtimer.py [hunk 8]
```diff
@@ -261,27 +346,30 @@ def main():
...
-        min_time = 'now-30m'
-        max_time = 'now'
-        refresh = '5s'
 
-    if not args.buckets:
+    if not args.buckets and not cbbackupmgr_stats_mode:
         buckets = stats_sources[0].get_buckets()
-    else:
+    elif args.buckets:
         buckets = sorted(args.buckets.split(','))
```
[the remark this answers is not in the document]
Comment (Lackshan) [reply promtimer/promtimer.py:362]: This is handled above, so we don't need to worry about buckets being None

File: README.md (hunk 7 above)
Comment (Lackshan) [reply README.md:132]: Forgot this, my bad

File: .gitmodules [hunk 9]
```diff
@@ -0,0 +1,3 @@
+[submodule "backup"]
```
Comment (dave-finlay) [inline .gitmodules:1]: Not crazy about creating this hard dependency. For one thing promtimer is public and backup is private. But also it the dependency seems unnecessary. I presume the only actual dependency is on cbmstatparser. I think this can just be expressed in the docs - that one needs to have cbmstatparser available on the path when running against a backup archive. Suggest remove the submodule.

File: promtimer/promtimer.py [hunk 10]
```diff
@@ -252,7 +287,57 @@ def main():
     grafana_port = args.grafana_port
     prometheus_base_port = grafana_port + 1
 
-    if not args.cluster:
+    cbbackupmgr_stats_mode = False
```
Comment (dave-finlay) [inline promtimer/promtimer.py:290]: This section is too long to be added inline. In fact, suggest adding a new `backupstats` module and adding this code there.

Comment (dave-finlay) [inline promtimer/promtimer.py:290]: Suggest calling this: `backup_archive_mode` to be consistent with the name you gave to the command line parameter.

**Review by dave-finlay — COMMENTED** [review]
will take another look after you get a chance to respond to my initial comments

**Comment by mattyhall** [comment]
@Lackshan there are a couple of places here where we are too strict about the format of the stat filenames, similar to https://issues.couchbase.com/browse/MB-56633. I've attached a diff that fixes it, but it's just what I bodge together - hopefully you can neaten it up :) 

[cbm_bodge.diff.zip](https://github.com/couchbaselabs/promtimer/files/11334378/cbm_bodge.diff.zip)

**Comment by Lackshan** [comment]
@dave-finlay Thanks for your review! I've addressed your comments and run my changes by @jameseh96 who's happy with them. Would appreciate it if you could take another look at it.

@mattyhall I've changed the stat file name logic to look for the first `-` from the right hand side so this should no longer be an issue.

**Comment by dave-finlay** [comment]
Hey Lackshan - 

This is too many minor commit to merge to master for what is essentially one idea (the addition of a new type of stats source and an ability to build a dashboard against it.) Please squash all of your commits into 1 and add a nice lengthy comment explaining clearly what it is your doing. Also: don't refer to the MB as it's not helpful. Simply explain the changes that you are making. 

I have a bunch more feedback on this PR which is too big and too complicated to add to the review comments here. Essentially, the problem is that this patch breaks regular promtimer (i.e. promtimer against a collection of normal cbcollects) and addressing this requires a bunch of work. In future, please make sure that you are not breaking the existing use cases when you're adding new functionality to something. In any case, it's easiest for me to push an alternate PR with the changes I think you need on top of your patches. It would be good if you can merge my changes into your PR and then update your PR with all the changes needed: https://github.com/couchbaselabs/promtimer/pull/63.

Will send you a quick email too.

**Comment by Lackshan** [comment]
Hey Dave, thanks for the review. Sorry about breaking the existing functionality, I've tested my changes again and squashed the commits down as you requested. I'll ask @couchbaseEd & @almidds99 to take a look.

**Review by almidds99 — APPROVED** [review]
+1

**Review by couchbaseEd — APPROVED** [review]
+1

**Comment by Lackshan** [comment]
@dave-finlay Would appreciate a review from you as the Support Tooling team are happy with this PR.

**Comment by dave-finlay** [comment]
Thanks Lackshan - appreciate the patch. It's probably going to take me a day or two to get back up-to-speed on the patch. But I have it on my todo list!

File: promtimer/promtimer.py [hunk 11]
```diff
@@ -100,17 +100,25 @@ def make_dashboards_yaml():
...
 
 
-def make_dashboards(stats_sources, buckets, min_time_string, max_time_string, refresh_string):
+def make_dashboards(
+    stats_sources,
+    buckets,
+    min_time_string,
+    max_time_string,
+    refresh_string,
+    dashboard_name_predicate):
```
Comment (dave-finlay) [inline promtimer/promtimer.py:109]: Can you make the indentation look better? I prefer this:
```
def make_dashboards(stats_sources,
                    buckets,
                    min_time_string,
                    max_time_string,
                    refresh_string,
                    dashboard_name_predicate):
```

File: promtimer/promtimer.py [hunk 12]
```diff
@@ -309,15 +326,15 @@ def main():
...
-        stats_sources = cbstats.CBCollect.get_stats_sources(prometheus_base_port)
-        if not stats_sources:
-            sys.exit(1)
-        times = cbstats.CBCollect.compute_min_and_max_times(stats_sources)
-        min_time = datetime.datetime.fromtimestamp(times[0]).isoformat()
-        max_time = datetime.datetime.fromtimestamp(times[1]).isoformat()
-        refresh = ''
-    else:
+    if live_cluster and args.backup_archive_path:
+        print("--cluster and --nodes are mutually exclusive ...")
```
Comment (dave-finlay) [inline promtimer/promtimer.py:330]: Change to "--cluster and --backup-archive-path are mutually exclusive options".

File: promtimer/promtimer.py [hunk 13]
```diff
@@ -309,15 +326,15 @@ def main():
...
+        sys.exit(1)
+
+    backup_archive_mode = False
```
Comment (dave-finlay) [inline promtimer/promtimer.py:333]: This is not needed as `backup_archive_mode` is always equal to `args.backup_archive_path` when compared as a boolean. Please remove.

File: promtimer/promtimer.py [hunk 14]
```diff
@@ -338,19 +355,40 @@ def main():
...
+        stats_sources = cbstats.CBCollect.get_stats_sources(prometheus_base_port)
+        if not stats_sources:
+            sys.exit(1)
+        times = cbstats.CBCollect.compute_min_and_max_times(stats_sources)
+        min_time = datetime.datetime.fromtimestamp(times[0]).isoformat()
+        max_time = datetime.datetime.fromtimestamp(times[1]).isoformat()
+        refresh = ''
 
-    if not args.buckets:
+
```
Comment (dave-finlay) [inline promtimer/promtimer.py:373]: Too many blank lines before the if statement (should just be one). Again see PEP standard.

**Review by dave-finlay — CHANGES_REQUESTED** [review]
Hi Lackshan. We are close. Can you address each of the comments and submit a follow-up patch with changes. Then we should basically be there.

File: promtimer/promtimer.py (hunk 13 above)
Comment (Lackshan) [reply promtimer/promtimer.py:333]: That's a good point. Not sure how this slipped by me 😞

File: README.md [hunk 15]
```diff
@@ -78,10 +78,19 @@ To get Promtimer, clone this repo locally:
...
-have them. 
+have them.
+
+### cbbackupmgr stats files
+For visualisation of cbbackupmgr stats files, you'll also need access to the private
+`backup` repository. You'll need to add the cbmstatparser binary to your path:
+```bash
+cd ~/backup
+go install ./cmd/cbmstatparser
+export PATH=$PATH:/backup/build/bin/cbmstatparser
```
[the remark this answers is not in the document]
Comment (Lackshan) [reply README.md:89]: `go install` places the executable [in `$GOPATH/pkg`](https://stackoverflow.com/a/30612612) which is meant to be in your PATH which means the user wouldn't have to add anything to their PATH or specify the location of the `cbmstatparser` executable manually.

My including the instructions to add the PATH here was to explain that users could just use the prebuilt version of `cbmstatparser` that is included in the backup repository.

That being said, I think you're right about adding the `--cbmstatparser-path` argument to promtimer. This is the most user-friendly option.

**Comment by Lackshan** [comment]
Thanks for the review @dave-finlay. I've made all the requested changes including adding the optional `--cbmstatparser-path` flag. I've also taken the liberty of: 
- Changing all copyright notices to say `2020-Present` 
- Correcting the number of newlines between functions to comply with PEP8 
- Adding vscode files into the `.gitignore` for convenience.

Would appreciate it if you could take another look at it again.

**Review by dave-finlay — CHANGES_REQUESTED** [review]
Just a few comments inline.

Also: can you remove the MB from the patch comment and expand it to briefly discuss what you are doing which is:
* adding a new type of stats source that can generate Prometheus metrics from backup archives
* starting Promtimer with a different out-of-the-box dashboard with panels built off these metrics
* that this use of Promtimer can be invoked with `--backup-archive-path`

The reference to the MB isn't helpful. Much better to simply put what you are doing and why in the patch comment.

File: promtimer/util.py [hunk 16]
```diff
@@ -306,3 +306,20 @@ def read_last_n_lines(filename, line_count=1):
...
+            while f.read(1) != b'\n':  # Read 1 byte
+                f.seek(-2, os.SEEK_CUR)
+        except OSError:
+            f.seek(0)
+
+        last_line = f.readline().decode()
+        if not last_line:
+            raise ValueError(f'CPU stats file at \'{filepath}\' is empty!')
+
+        return last_line.split(';')[0]
```
[the remark this answers is not in the document]
Comment (Lackshan) [reply promtimer/util.py:325]: This is a good catch. I've done this and will ensure I make methods like these generic going forwards.

**Comment by Lackshan** [comment]
@dave-finlay I've addressed these comments and improved both the commit message and name of this PR as suggested. Would appreciate another look from you. Thanks!

**Comment by dave-finlay** [comment]
Looks good. Will merge. Thanks Lackshan!
